Address Rails 5.1 changes to ActiveModel::Dirty

johnnaegle opened the pull request:
In Rails 5.1, new methods were added that split dirty attribute tracking into _before_ save and _after_ save versions.  Before save, you have will_save_change_to_<attribute>? and in after save you have saved_change_to_<attribute>?, there are a lot of these methods and in Rails 5.2 the behavior of accessing a <attribute>_changed? method or <record>_changed? in an after callback is changed to be the opposite behavior of what it is now.  Rails 5.1 added deprecation warnings for these.

- replace calls to activemodel defined <attribute>_changed? and <attribute>_was?  methods with will_save_change_to_<attribute>? and <attribute_in_database>

Commit in Rails:
https://github.com/rails/rails/commit/16ae3db5a5c6a08383b974ae6c96faac5b4a3c81

jaredbeck commented:
Hi John, thanks for the contribution!

- What problem were you trying to fix here? Deprecation warnings?
- Is it possible to add a test for this?
- Please add an entry to the changelog under 5.0.0 -> Fixed

johnnaegle commented:
Hi Jared,

I added a test and updated the changelog., As this its fixing a deprecation warning in Rails 5.1, it is not the most straight forward thing to see with the master branch.    I did the following to see the depreciation warning:

- Add back support for 5.1 (gemspec, gemfiles/rails-5.1.rb)
- git checkout HEAD~1
- BUNDLE_GEMFILE=gemfiles/rails-5.1.rb bundle install
- BUNDLE_GEMFILE=gemfiles/rails-5.1.rb bundle exec rake 2>&1 | grep "DEPRECATION WARNING"
- git checkout active-record-dirty
- BUNDLE_GEMFILE=gemfiles/rails-5.1.rb bundle exec rake 2>&1 | grep "DEPRECATION WARNING"

jaredbeck commented:
Thanks John. Are you only trying to fix deprecation warnings in 5.1, or do we need to change which methods we're using anyway?

jaredbeck commented on lib/authlogic/acts_as_authentic/perishable_token.rb:
```diff
@@ -58,7 +58,7 @@ def self.included(klass)
             include InstanceMethods
 
-            validates_uniqueness_of :perishable_token, if: :perishable_token_changed?
+            validates_uniqueness_of :perishable_token, if: :will_save_change_to_perishable_token?
```
This change (in a `before_save`) is not strictly necessary, but is recommended for clarity?

jaredbeck commented:
I've pushed some suggested changes (https://github.com/binarylogic/authlogic/pull/638/commits/976c975afcf622eeef3a019b7fc3051fec7a64f3) which I think simplify the new test. Please take a look.

johnnaegle commented:
Thanks! Sorry for going silent

> I've pushed some suggested changes ([976c975](https://github.com/binarylogic/authlogic/commit/976c975afcf622eeef3a019b7fc3051fec7a64f3)) which I think simplify the new test. Please take a look.

Looks good, thanks for reviewing and merging
